### PR TITLE
KZOO-308 Functionality to Intercept Users According to CALEA Intercept Requests

### PR DESCRIPTION
**Pull request by nkarunarathne**
Previously, there was no option to tap users according to CALEA intercept requests.

This change will allow wiretap recordings to save along with other recording types without impacting existing functionality.

### CODE REVIEW

File: src/media_recording.erl [hunk 1]
```diff
@@ -386,7 +398,16 @@ check_url_param(Param) ->
...
+save_recording(#{media := Media, pid := Pid, event:= Event}=Store) ->
+    case kz_json:get_value([<<"Recording">>, <<"Is-Wiretap">>], Event, <<"false">>) of
+        <<"false">> ->
+            UpdatedStore = Store;
+        <<"true">> ->
+            UpdatedStore = update_wiretap_store_info(Store)
+    end,
+    maybe_store_recording(Media, UpdatedStore, Pid).
+
+maybe_store_recording(Media, Store, Pid) ->
```
Comment (icehess) [inline src/media_recording.erl]: We don't like binding variables this way in different case clasues. Instead update store in a separate function like this:


```erlang
save_recording(#{media := Media, pid := Pid}=Store) ->
    UpdatedStore = maybe_add_wiretap_info(Store),
    case maybe_store_recording_meta(UpdatedStore) of
        {'error', Err} ->
            lager:warning("error storing metadata : ~p", [Err]),
            gen_server:cast(Pid, {'store_failed', UpdatedStore});
        Rev ->
            StoreUrl = fun()-> store_url(UpdatedStore, Rev) end,
            store_recording(Media, StoreUrl, UpdatedStore)
    end.

-spec maybe_add_wiretap_info(store_info()) -> store_info().
maybe_add_wiretap_info(#{event:= Event}=Store) ->
    case kz_json:is_true([<<"Recording">>, <<"Is-Wiretap">>], Event, 'false') of
        'false' -> Store;
        'true' -> update_wiretap_store_info(Store)
    end.

```

File: src/media_recording.erl [hunk 2]
```diff
@@ -467,3 +488,49 @@ maybe_set_default_user_metadata(#{account_id := AccountId}=StoreInfo) ->
...
+                        ,retries => ?WIRETAP_STORAGE_RETRY_TIMES
+                        ,verb => Verb
+                        ,account_id => AccountId
+                        ,endpoint_id => EndpointId
+                        ,call_id => CallId
+                        ,event => JObj
+                        ,origin => Origin
+                        ,pid => Pid
+                        },
+    UpdatedStoreInfo.
```
Comment (icehess) [inline src/media_recording.erl]: You don't need to match on all maps keys and re-add them, just binding to var and update the only the fields you want to update:

```suggestion
-spec update_wiretap_store_info(store_info()) -> store_info().
update_wiretap_store_info(Store) ->
    Store#{doc_db => ?WIRETAP_DB
          ,should_store => get_should_store()
          ,retries => ?WIRETAP_STORAGE_RETRY_TIMES
          }.
```

File: src/media_recording.erl [hunk 3]
```diff
@@ -467,3 +488,49 @@ maybe_set_default_user_metadata(#{account_id := AccountId}=StoreInfo) ->
...
+
+get_should_store() ->
+    case kapps_config:get(?CONFIG_CAT ,<<"wiretap_should_store">>, <<"local">>) of
```
Comment (icehess) [inline src/media_recording.erl]:
```suggestion
    case kapps_config:get_ne_binary(?CONFIG_CAT ,<<"wiretap_should_store">>, <<"local">>) of
```

**Review by icehess — CHANGES_REQUESTED** [review]

**Comment by nkarunarathne** [comment]
Not relevant now

**Pull request closed without merging**